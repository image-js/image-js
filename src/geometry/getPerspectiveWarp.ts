import { Matrix, SingularValueDecomposition } from 'ml-matrix';

import type { Point } from '../utils/geometry/points.js';

/**
 * Options for getPerspectiveWarp function.
 */
export interface GetPerspectiveWarpOptions {
  /**
   * The horizontal dimension (in pixels) of the final rectified rectangular image.
   */
  width?: number;
  /**
   * The vertical dimension (in pixels) of the final rectified rectangular image.
   */
  height?: number;
}
/**
 * Returns result matrix along with vertical and horizontal dimensions for the rectangular image.
 */
export type GetPerspectiveWarpData = Required<GetPerspectiveWarpOptions> & {
  matrix: number[][];
};

// REFERENCES :
// http://graphics.cs.cmu.edu/courses/15-463/2011_fall/Lectures/morphing.pdf
/**
 * Computes a perspective transformation matrix to rectify a quadrilateral region into a rectangle.
 *
 * This function takes four corner points of a quadrilateral (e.g., a document photographed at an angle) and calculates the transformation matrix needed to "unwarp" it into a rectangular image.
 * The output dimensions can be specified or calculated automatically based on the input geometry.
 * @param pts - 4 reference corners.
 * @param options - PerspectiveWarpOptions.
 * @returns - Matrix from 4 points.
 */
export function getPerspectiveWarp(
  pts: Point[],
  options: GetPerspectiveWarpOptions = {},
): GetPerspectiveWarpData {
  if (pts.length !== 4) {
    throw new Error(
      `The array pts must have four elements, which are the four corners. Currently, pts have ${pts.length} elements`,
    );
  }
  const { width, height } = options;
  const [tl, tr, br, bl] = order4Points(pts);

  let widthRect;
  let heightRect;
  if (height && width) {
    widthRect = width;
    heightRect = height;
  } else if (!height && !width) {
    widthRect = Math.ceil(
      Math.max(distance2Points(tl, tr), distance2Points(bl, br)),
    );
    heightRect = Math.ceil(
      Math.max(distance2Points(tl, bl), distance2Points(tr, br)),
    );
  } else {
    throw new Error(
      `Invalid dimensions: ${width ? '`height`' : '`width`'} is missing. ` +
        `Either provide both width and height, or omit both to auto-calculate dimensions.`,
    );
  }

  const [x1, y1] = [0, 0];
  const [x2, y2] = [widthRect - 1, 0];
  const [x3, y3] = [widthRect - 1, heightRect - 1];
  const [x4, y4] = [0, heightRect - 1];

  const S = new Matrix([
    [x1, y1, 1, 0, 0, 0, -x1 * tl.column, -y1 * tl.column],
    [x2, y2, 1, 0, 0, 0, -x2 * tr.column, -y2 * tr.column],
    [x3, y3, 1, 0, 0, 0, -x3 * br.column, -y3 * br.column],
    [x4, y4, 1, 0, 0, 0, -x4 * bl.column, -y4 * bl.column],
    [0, 0, 0, x1, y1, 1, -x1 * tl.row, -y1 * tl.row],
    [0, 0, 0, x2, y2, 1, -x2 * tr.row, -y2 * tr.row],
    [0, 0, 0, x3, y3, 1, -x3 * br.row, -y3 * br.row],
    [0, 0, 0, x4, y4, 1, -x4 * bl.row, -y4 * bl.row],
  ]);
  const D = Matrix.columnVector([
    tl.column,
    tr.column,
    br.column,
    bl.column,
    tl.row,
    tr.row,
    br.row,
    bl.row,
  ]);

  const svd = new SingularValueDecomposition(S);
  const T = svd.solve(D).to1DArray(); // solve S*T = D
  T.push(1);

  const M = [];
  for (let i = 0; i < 3; i++) {
    const row = [];
    for (let j = 0; j < 3; j++) {
      row.push(T[i * 3 + j]);
    }
    M.push(row);
  }
  return { matrix: M, width: widthRect, height: heightRect };
}

/**
 * Sorts 4 points in order =>[top-left,top-right,bottom-right,bottom-left]. Input points must be in clockwise or counter-clockwise order.
 * @param pts - Array of 4 points.
 * @returns Sorted array of 4 points.
 */
export function order4Points(pts: Point[]) {
  let tl: Point;
  let tr: Point;
  let br: Point;
  let bl: Point;

  let minX = pts[0].column;
  let indexMinX = 0;

  for (let i = 1; i < pts.length; i++) {
    if (pts[i].column < minX) {
      minX = pts[i].column;
      indexMinX = i;
    }
  }

  let minX2 = pts[(indexMinX + 1) % pts.length].column;
  let indexMinX2 = (indexMinX + 1) % pts.length;

  for (let i = 0; i < pts.length; i++) {
    if (pts[i].column < minX2 && i !== indexMinX) {
      minX2 = pts[i].column;
      indexMinX2 = i;
    }
  }
  if (pts[indexMinX2].row < pts[indexMinX].row) {
    tl = pts[indexMinX2];
    bl = pts[indexMinX];
    if (indexMinX !== (indexMinX2 + 1) % 4) {
      tr = pts[(indexMinX2 + 1) % 4];
      br = pts[(indexMinX2 + 2) % 4];
    } else {
      tr = pts[(indexMinX2 + 2) % 4];
      br = pts[(indexMinX2 + 3) % 4];
    }
  } else {
    bl = pts[indexMinX2];
    tl = pts[indexMinX];
    if (indexMinX2 !== (indexMinX + 1) % 4) {
      tr = pts[(indexMinX + 1) % 4];
      br = pts[(indexMinX + 2) % 4];
    } else {
      tr = pts[(indexMinX + 2) % 4];
      br = pts[(indexMinX + 3) % 4];
    }
  }
  return [tl, tr, br, bl];
}
/**
 *  Calculates distance between points.
 * @param p1 - Point1
 * @param p2 - Point2
 * @returns distance between points.
 */
function distance2Points(p1: Point, p2: Point) {
  return Math.hypot(p1.column - p2.column, p1.row - p2.row);
}
