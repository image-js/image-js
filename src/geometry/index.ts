export * from './resize.js';
export * from './rotate.js';
export * from './transform.js';
export * from './transformRotate.js';
export type { Point } from '../utils/geometry/points.js';
