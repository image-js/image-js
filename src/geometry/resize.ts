import type { Image } from '../Image.js';
import type { BorderType } from '../utils/interpolateBorder.js';
import type { InterpolationType } from '../utils/interpolatePixel.js';
import { assert } from '../utils/validators/assert.js';

import { transform } from './transform.js';

export interface ResizeOptions {
  /**
   * Width of the output image.
   */
  width?: number;
  /**
   * Height of the output image.
   */
  height?: number;
  /**
   * Factor by which to scale the width.
   */
  xFactor?: number;
  /**
   * Factor by which to scale the width.
   */
  yFactor?: number;
  /**
   * Whether the aspect ratio of the image should be preserved.
   * @default `true`
   */
  preserveAspectRatio?: boolean;
  /**
   * Method to use to interpolate the new pixels.
   * @default `'bilinear'`
   */
  interpolationType?: InterpolationType;
  /**
   * Specify how the borders should be handled.
   * @default `'constant'`
   */
  borderType?: BorderType;
  /**
   * Value of the border if BorderType is 'constant'.
   * @default `0`
   */
  borderValue?: number;
}

/**
 * Returns a resized copy of an image.
 * @param image - Original image.
 * @param options - Resize options.
 * @returns The new image.
 */
export function resize(image: Image, options: ResizeOptions): Image {
  const {
    interpolationType = 'bilinear',
    borderType = 'replicate',
    borderValue = 0,
  } = options;
  const { width, height, xFactor, yFactor } = checkOptions(image, options);

  return transform(
    image,
    [
      [xFactor, 0, xFactor / 2],
      [0, yFactor, yFactor / 2],
    ],
    {
      interpolationType,
      borderType,
      borderValue,
      height,
      width,
    },
  );
}

/**
 * Verify that the resize options are valid.
 * @param image - Image.
 * @param options - Resize options.
 * @returns Resize options.
 */
function checkOptions(
  image: Image,
  options: ResizeOptions,
): { width: number; height: number; xFactor: number; yFactor: number } {
  const {
    width,
    height,
    xFactor,
    yFactor,
    preserveAspectRatio = true,
  } = options;

  if (
    width === undefined &&
    height === undefined &&
    xFactor === undefined &&
    yFactor === undefined
  ) {
    throw new TypeError(
      'at least one of the width, height, xFactor or yFactor options must be passed',
    );
  }

  let newWidth: number;
  let newHeight: number;

  const maybeWidth = getSize(width, xFactor, image.width, preserveAspectRatio);

  const maybeHeight = getSize(
    height,
    yFactor,
    image.height,
    preserveAspectRatio,
  );

  if (maybeWidth === undefined) {
    assert(maybeHeight !== undefined);
    newWidth = Math.round(maybeHeight * (image.width / image.height));
  } else {
    newWidth = maybeWidth;
  }

  if (maybeHeight === undefined) {
    assert(maybeWidth !== undefined);
    newHeight = Math.round(maybeWidth * (image.height / image.width));
  } else {
    newHeight = maybeHeight;
  }

  return {
    width: newWidth,
    height: newHeight,
    xFactor: xFactor ?? newWidth / image.width,
    yFactor: yFactor ?? newHeight / image.height,
  };
}

/**
 * Compute automatic new size.
 * @param sizeOpt - Size option.
 * @param factor - Factor option.
 * @param sizeImg - Size of the image.
 * @param preserveAspectRatio - Whether to preserve the aspect ratio.
 * @returns New size.
 */
function getSize(
  sizeOpt: number | undefined,
  factor: number | undefined,
  sizeImg: number,
  preserveAspectRatio: boolean,
): number | undefined {
  if (sizeOpt === undefined) {
    if (factor !== undefined) {
      return Math.round(sizeImg * factor);
    } else if (!preserveAspectRatio) {
      return sizeImg;
    }
  } else if (factor !== undefined) {
    throw new TypeError('factor and size cannot be passed together');
  } else {
    return sizeOpt;
  }
  return undefined;
}
