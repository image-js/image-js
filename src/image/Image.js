import bitMethods from './core/bitMethods';
import checkProcessable from './core/checkProcessable';
import exportMethods from './core/export';
import { extendMethod, extendProperty } from './core/extend';
import getRGBAData from './core/getRGBAData';
import {
  getKind,
  verifyKindDefinition,
  createPixelArray,
  getTheoreticalPixelArraySize,
} from './core/kind';
import { RGBA } from './core/kindNames';
import load from './core/load';
import valueMethods from './core/valueMethods';
import extend from './extend';
import getImageParameters from './internal/getImageParameters';
import RoiManager from './roi/manager';

const objectToString = Object.prototype.toString;

/**
 * Class representing an image.
 * This class allows to manipulate easily images directly in the browser or in node.
 *
 * This library is designed to deal with scientific images (8 or 16 bit depth) and will be able to open
 * and process jpeg, png and uncompressed tiff images. It is designed to work in the browser
 * as on the server side in node.
 *
 * An image is characterized by:
 * * width and height
 * * colorModel (RGB, HSL, CMYK, GREY, ...)
 * * components: number of components, Grey scale images will have 1 component while RGB will have 3 and CMYK 4.
 * * alpha: 0 or 1 depending if there is an alpha channel. The
 *      alpha channel define the opacity of each pixel
 * * channels: number of channels (components + alpha)
 * * bitDepth : number of bits to define the intensity of a point.
 *      The values may be 1 for a binary image (mask), 8 for a normal image (each
 *      channel contains values between 0 and 255) and 16 for scientific images
 *      (each channel contains values between 0 and 65535).
 *      The png library and tiff library included in image-js allow to deal correctly with
 *      8 and 16 bit depth images.
 * * position : an array of 2 elements that allows to define a relative position
 *      to a parent image. This will be used in a crop or in the management
 *      of Region Of Interests (Roi) for example
 * * data : an array that contains all the points of the image.
 *      Depending the bitDepth Uint8Array (1 bit), Uint8Array (8 bits),
 *      Uint16Array (16 bits), Float32Array (32 bits)
 *
 * In an image there are pixels and points:
 * * A pixel is an array that has as size the number of channels
 * and that contains all the values that define a particular pixel of the image.
 * * A point is an array of 2 elements that contains the x / y coordinate
 * of a specific pixel of the image
 *
 *
 * @class Image
 * @param {number} [width=1]
 * @param {number} [height=1]
 * @param {Array} [data] - Image data to load
 * @param {object} [options]
 *
 *
 * @example
 * // JavaScript code using Node.js to get some info about the image.
 * // We load the library that was installed using 'npm install image-js'
 * const { Image } = require('image-js');
 *
 * // Loading an image is asynchronous and will return a Promise.
 * Image.load('cat.jpg').then(function (image) {
 *   console.log('Width', image.width);
 *   console.log('Height', image.height);
 *   console.log('colorModel', image.colorModel);
 *   console.log('components', image.components);
 *   console.log('alpha', image.alpha);
 *   console.log('channels', image.channels);
 *   console.log('bitDepth', image.bitDepth);
 * });
 *
 * @example
 * // Convert an image to greyscale
 * const { Image } = require('image-js');
 *
 * Image.load('cat.jpg').then(function (image) {
 *   var grey = image.grey();
 *   grey.save('cat-grey.jpg');
 * });
 *
 * @example
 * // Split an RGB image in its components
 * const { Image } = require('image-js');
 *
 * Image.load('cat.jpg').then(function (image) {
 *   var components = image.split();
 *   components[0].save('cat-red.jpg');
 *   components[1].save('cat-green.jpg');
 *   components[2].save('cat-blur.jpg');
 * });
 *
 *
 * @example
 * // For this example you will need the picture of an ecstasy pill that is available on
 * // wget https://raw.githubusercontent.com/image-js/core/854e70f50d63cc73d2dde1d2020fe61ba1b5ec05/test/img/xtc.png // the goal is to isolate the picture and to get a RGB histogram of the pill.
 * // Practically this allows to classify pills based on the histogram similarity
 * // This work was published at: http://dx.doi.org/10.1016/j.forsciint.2012.10.004
 *
 * const { Image } = require('image-js');
 *
 * const image = await Image.load('xtc.png');
 *
 * const grey = image.grey({
 *   algorithm:'lightness'
 * });
 * // we create a mask, which is basically a binary image
 * // a mask has as source a grey image and we will decide how to determine
 * // the threshold to define what is white and what is black
 * var mask = grey.mask({
 *   algorithm: 'li'
 * });
 *
 * // it is possible to create an array of Region Of Interest (Roi) using
 * // the RoiManager. A RoiManager will be applied on the original image
 * // in order to be able to extract from the original image the regions
 *
 * // the result of this console.log result can directly be pasted
 * // in the browser
 * // console.log(mask.toDataURL());
 *
 *
 * var manager = image.getRoiManager();
 * manager.fromMask(mask);
 * var rois = manager.getRoi({
 *   positive: true,
 *   negative: false,
 *   minSurface: 100
 * });
 *
 * // console.log(rois);
 *
 * // we can sort teh rois by surface
 * // for demonstration we use an arrow function
 * rois.sort((a, b) => b.surface - a.surface);
 *
 * // the first Roi (the biggest is expected to be the pill)
 *
 * var pillMask = rois[0].getMask({
 *   scale: 0.7   // we will scale down the mask to take just the center of the pill and avoid border effects
 * });
 *
 * // image-js remembers the parent of the image and the relative
 * // position of a derived image. This is the case for a crop as
 * // well as for Roi
 *
 * var pill = image.extract(pillMask);
 * pill.save('pill.jpg');
 *
 * var histogram = pill.getHistograms({ maxSlots: 16 });
 *
 * console.log(histogram);
 *
 * @example
 * // Example of use of IJS in the browser
 *
 * <script>
 *  var canvas = document.getElementById('myCanvasID');
 *  var image = IJS.fromCanvas(canvas);
 * </script>
 *
 * @example
 * // Image from matrix of values
 * const [min, max] = d3.extent(temperatures)
 * const colorScaler = d3.scaleSequential([min, max], d3.interpolateRdYlBu);
 *
 * // size = rows * columns * channels
 * const data = new Uint8Array(2*3*3);
 * for (let i = 0; i < temperatures.length; i++) {
 *   const {r, g, b} = d3.rgb(colorScaler(temperatures[i]));
 *   data[i*3] = r;
 *   data[i*3 + 1] = g;
 *   data[i*3 + 2] = b;
 * }
 *
 * const image = new Image(2, 3, data, { kind: 'RGB' });
 * // or
 * const image = new Image({ width: 2, height: 3, data, kind: 'RGB'});
 */
export default class Image {
  constructor(width, height, data, options) {
    if (arguments.length === 1) {
      options = width;
      ({ width, height, data } = options);
    } else if (data && !data.length) {
      options = data;
      ({ data } = options);
    }
    if (width === undefined) width = 1;
    if (height === undefined) height = 1;
    if (options === undefined) options = {};

    if (typeof options !== 'object' || options === null) {
      throw new TypeError('options must be an object');
    }

    if (!Number.isInteger(width) || width <= 0) {
      throw new RangeError('width must be a positive integer');
    }
    if (!Number.isInteger(height) || height <= 0) {
      throw new RangeError('height must be a positive integer');
    }

    const { kind = RGBA } = options;
    if (typeof kind !== 'string') {
      throw new TypeError('kind must be a string');
    }
    const theKind = getKind(kind);
    const kindDefinition = Object.assign({}, options);
    for (const prop in theKind) {
      if (kindDefinition[prop] === undefined) {
        kindDefinition[prop] = theKind[prop];
      }
    }
    verifyKindDefinition(kindDefinition);

    const { components, bitDepth, colorModel } = kindDefinition;
    const alpha = kindDefinition.alpha + 0;
    const size = width * height;
    const channels = components + alpha;
    const maxValue = bitDepth === 32 ? Number.MAX_VALUE : 2 ** bitDepth - 1;

    if (data === undefined) {
      data = createPixelArray(
        size,
        components,
        alpha,
        channels,
        bitDepth,
        maxValue,
      );
    } else {
      const expectedLength = getTheoreticalPixelArraySize(
        size,
        channels,
        bitDepth,
      );
      if (data.length !== expectedLength) {
        throw new RangeError(
          `incorrect data size: ${data.length}. Should be ${expectedLength}`,
        );
      }

      if (bitDepth === 1) {
        const exceedingBits = data.length * 8 - size * channels;
        data[data.length - 1] =
          data[data.length - 1] & ~(0xff >> (8 - exceedingBits));
      }
    }

    /**
     * Width of the image.
     * @member {number}
     */
    this.width = width;

    /**
     * Height of the image.
     * @member {number}
     */
    this.height = height;

    /**
     * Typed array holding the image data.
     * @member {TypedArray}
     */
    this.data = data;

    /**
     * Total number of pixels (width * height).
     * @member {number}
     */
    this.size = size;

    /**
     * Number of color channels in the image.
     * A grey image has 1 component. An RGB image has 3 components.
     * @member {number}
     */
    this.components = components;

    /**
     * Alpha is 1 if there is an alpha channel, 0 otherwise.
     * @member {number}
     */
    this.alpha = alpha;

    /**
     * Number of bits per value in each channel.
     * @member {number}
     */
    this.bitDepth = bitDepth;

    /**
     * Maximum value that a pixel can have.
     * @member {number}
     */
    this.maxValue = maxValue;

    /**
     * Color model of the image.
     * @member {ColorModel}
     */
    this.colorModel = colorModel;

    /**
     * Total number of channels. Is equal to `image.components + image.alpha`.
     * @member {number}
     */
    this.channels = channels;

    /**
     * Metadata associated with the image.
     * @member {object}
     */
    this.meta = options.meta || {};

    // TODO review those props
    Object.defineProperty(this, 'parent', {
      enumerable: false,
      writable: true,
      configurable: true,
      value: options.parent || null,
    });
    this.position = options.position || [0, 0];

    this.computed = null;
    this.sizes = [this.width, this.height];
    this.multiplierX = this.channels;
    this.multiplierY = this.channels * this.width;
    this.isClamped = this.bitDepth < 32;
    this.borderSizes = [0, 0]; // when a filter creates a border, it may have impact on future processing like Roi
  }

  get [Symbol.toStringTag]() {
    return 'IJSImage';
  }

  static isImage(object) {
    return objectToString.call(object) === '[object IJSImage]';
  }

  /**
   * Creates an image from an HTML Canvas object
   * @param {Canvas} canvas
   * @return {Image}
   */
  static fromCanvas(canvas) {
    const ctx = canvas.getContext('2d');
    const imageData = ctx.getImageData(0, 0, canvas.width, canvas.height);
    return new Image(imageData.width, imageData.height, imageData.data);
  }

  /**
   * Create a new Image based on the characteristics of another one.
   * @param {Image} other
   * @param {object} [options] - Override options to change some parameters
   * @return {Image}
   * @example
   * const newImage = Image.createFrom(image, { width: 100 });
   */
  static createFrom(other, options) {
    const newOptions = getImageParameters(other);
    Object.assign(
      newOptions,
      {
        parent: other,
        position: [0, 0],
      },
      options,
    );
    return new Image(newOptions);
  }

  /**
   * Create a new manager for regions of interest based on the current image.
   * @param {object} [options]
   * @return {RoiManager}
   */
  getRoiManager(options) {
    return new RoiManager(this, options);
  }

  /**
   * Create a copy a the current image, including its data.
   * @instance
   * @return {Image}
   */
  clone() {
    const newData = this.data.slice();
    return new Image(this.width, this.height, newData, this);
  }

  apply(filter) {
    for (let y = 0; y < this.height; y++) {
      for (let x = 0; x < this.width; x++) {
        let index = (y * this.width + x) * this.channels;
        filter.call(this, index);
      }
    }
  }
}

valueMethods(Image);
bitMethods(Image);
exportMethods(Image);

Image.prototype.checkProcessable = checkProcessable;
Image.prototype.getRGBAData = getRGBAData;

Image.load = load;
Image.extendMethod = extendMethod;
Image.extendProperty = extendProperty;
extend(Image);
