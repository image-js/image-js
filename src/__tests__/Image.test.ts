import { inspect } from 'node:util';

import { Image, ImageCoordinates } from '../Image';
import { Point } from '../geometry';

describe('create new images', () => {
  it('should create a 8-bit image', () => {
    const img = new Image(10, 20);
    expect(img).toMatchObject({
      width: 10,
      height: 20,
      size: 200,
      bitDepth: 8,
      colorModel: 'RGB',
      components: 3,
      channels: 3,
      alpha: false,
      maxValue: 255,
    });
    expect(img.getRawImage().data).toHaveLength(600);
  });

  it('should create a 16-bit image', () => {
    const img = new Image(10, 20, { bitDepth: 16 });
    expect(img).toMatchObject({
      width: 10,
      height: 20,
      size: 200,
      bitDepth: 16,
      colorModel: 'RGB',
      components: 3,
      channels: 3,
      alpha: false,
      maxValue: 65535,
    });
    expect(img.getRawImage().data).toHaveLength(600);
  });

  it('should create a grey image with alpha', () => {
    const img = new Image(10, 20, { colorModel: 'GREYA' });
    expect(img).toMatchObject({
      bitDepth: 8,
      colorModel: 'GREYA',
      components: 1,
      channels: 2,
      alpha: true,
    });
    expect(img.getRawImage().data).toHaveLength(400);
  });

  it('should create from existing data array', () => {
    // prettier-ignore
    const data = Uint8Array.from([
      0, 1, 2,
      3, 4, 5
    ]);
    const img = new Image(3, 2, { data, colorModel: 'GREY' });
    expect(img.getValue(0, 1, 0)).toBe(3);
    expect(img.getRawImage().data).toBe(data);
  });

  it('should throw on wrong width', () => {
    expect(() => new Image(0, 1)).toThrow(
      /width must be an integer and at least 1. Received 0/,
    );
    expect(() => new Image(0.5, 1)).toThrow(
      /width must be an integer and at least 1. Received 0.5/,
    );
  });

  it('should throw on wrong height', () => {
    expect(() => new Image(1, 0)).toThrow(
      /height must be an integer and at least 1. Received 0/,
    );
    expect(() => new Image(1, 0.5)).toThrow(
      /height must be an integer and at least 1. Received 0.5/,
    );
  });

  it('should throw on wrong data size', () => {
    const data = new Uint16Array(2);
    expect(() => new Image(2, 2, { data, bitDepth: 16 })).toThrow(
      /incorrect data size: 2. Expected 12/,
    );
  });

  it('should throw on wrong bit depth', () => {
    // @ts-expect-error we want to test the error.
    expect(() => new Image(1, 1, { bitDepth: 20 })).toThrow(
      /invalid bitDepth: 20/,
    );
  });

  it('should throw with bit depth 8 but data 16', () => {
    const data = new Uint16Array([1, 2, 3, 4]);
    expect(() => new Image(2, 2, { colorModel: 'GREY', data })).toThrow(
      'bitDepth is 8 but data is Uint16Array',
    );
  });
  it('should throw with bit depth 16 but data 8', () => {
    const data = new Uint8Array([1, 2, 3, 4]);
    expect(
      () =>
        new Image(2, 2, {
          colorModel: 'GREY',
          bitDepth: 16,
          data,
        }),
    ).toThrow('bitDepth is 16 but data is Uint8Array');
  });
});

describe('get and set pixels', () => {
  it('should get and set', () => {
    const img = new Image(10, 20);
    expect(img.getPixel(5, 15)).toStrictEqual([0, 0, 0]);
    img.setPixel(5, 15, [1, 3, 5]);
    expect(img.getPixel(5, 15)).toStrictEqual([1, 3, 5]);
  });

  it('should get and set by index', () => {
    const img = new Image(10, 20);
    expect(img.getPixelByIndex(5)).toStrictEqual([0, 0, 0]);
    img.setPixelByIndex(5, [1, 3, 5]);
    expect(img.getPixelByIndex(5)).toStrictEqual([1, 3, 5]);
  });

  it('get pixel by index in grey image', () => {
    const img = testUtils.createGreyImage([
      [1, 2, 3],
      [4, 5, 6],
    ]);
    expect(img.getPixelByIndex(0)).toStrictEqual([1]);
    expect(img.getPixelByIndex(4)).toStrictEqual([5]);
  });

  it('should get and set value', () => {
    const img = new Image(10, 20);
    expect(img.getValue(5, 15, 0)).toBe(0);
    img.setValue(5, 15, 0, 50);
    expect(img.getValue(5, 15, 0)).toBe(50);
  });

  it('should get and set value by index', () => {
    const img = new Image(10, 20);
    expect(img.getValueByIndex(15, 0)).toBe(0);
    img.setValueByIndex(15, 0, 50);
    expect(img.getValueByIndex(15, 0)).toBe(50);
  });

  it('should get and set value by point', () => {
    const point = { column: 15, row: 0 };
    const img = new Image(10, 20);
    expect(img.getValueByPoint(point, 0)).toBe(0);
    img.setValueByPoint(point, 0, 50);
    expect(img.getValueByPoint(point, 0)).toBe(50);
  });
});

test('createFrom', () => {
  const img = new Image(2, 20, { colorModel: 'GREY' });
  const newImg = Image.createFrom(img);
  expect(img.width).toBe(newImg.width);
  expect(img.height).toBe(newImg.height);
  expect(img.colorModel).toBe(newImg.colorModel);
  expect(img.origin).toStrictEqual(newImg.origin);
});

test('clone', () => {
  const img = new Image(2, 2, { colorModel: 'GREY' });
  img.setValue(1, 0, 0, 50);
  const copy = img.clone();
  expect(copy).toMatchImage(img);
});

test('changeEach', () => {
  const img = new Image(1, 2);
  let i = 0;
  img.changeEach(() => i++);
  expect(img).toMatchImageData([
    [0, 1, 2],
    [3, 4, 5],
  ]);
});

test.each<[ImageCoordinates, Point]>([
  ['bottom-left', { column: 0, row: 4 }],
  ['bottom-right', { column: 3, row: 4 }],
  ['center', { column: 1.5, row: 2 }],
  ['top-left', { column: 0, row: 0 }],
  ['top-right', { column: 3, row: 0 }],
])('getCoordinates - %s', (coordinates, point) => {
  const img = new Image(4, 5);
  expect(img.getCoordinates(coordinates)).toStrictEqual(point);
});

test('getCoordinates - with rounding', () => {
  const img = new Image(4, 5);
  expect(img.getCoordinates('center', true)).toStrictEqual({
    column: 2,
    row: 2,
  });
});

test('getCoordinates - bad parameter', () => {
  const img = new Image(4, 5);
  // @ts-expect-error bad parameter
  expect(() => img.getCoordinates('bad')).toThrow(
    /invalid image coordinates: bad/,
  );
});

test('fill with a constant color', () => {
  const img = new Image(2, 2);
  img.fill(50);
  expect(img).toMatchImageData([
    [50, 50, 50, 50, 50, 50],
    [50, 50, 50, 50, 50, 50],
  ]);
});

test('fill with a float value', () => {
  const img = new Image(2, 2);
  img.fill(10.7);
  expect(img).toMatchImageData([
    [10, 10, 10, 10, 10, 10],
    [10, 10, 10, 10, 10, 10],
  ]);
});

test('fill with a color as RGBA array', () => {
  const img = new Image(1, 2);
  img.fill([1, 2, 3]);
  expect(img).toMatchImageData([
    [1, 2, 3],
    [1, 2, 3],
  ]);
});

test('fill with out of range value', () => {
  const img = new Image(1, 1);
  expect(() => img.fill(256)).toThrow(
    /invalid value: 256. It must be a positive value smaller than 256/,
  );
});

test('fill with out of range value in array', () => {
  const img = new Image(1, 1);
  expect(() => img.fill([0, -1, 2])).toThrow(
    /invalid value: -1. It must be a positive value smaller than 256/,
  );
});

test('fill with channel mismatch', () => {
  const img = new Image(1, 1);
  expect(() => img.fill([0, 1, 2, 3])).toThrow(
    /the size of value must match the number of channels \(3\). Received 4/,
  );
});

test('fill channel 0', () => {
  const img = new Image(1, 2);
  img.fillChannel(0, 50);
  expect(img).toMatchImageData([
    [50, 0, 0],
    [50, 0, 0],
  ]);
});

test('fill channel 2', () => {
  const img = new Image(1, 2);
  img.fillChannel(2, 50);
  expect(img).toMatchImageData([
    [0, 0, 50],
    [0, 0, 50],
  ]);
});

test('fill channel invalid channel', () => {
  const img = new Image(1, 2);
  expect(() => img.fillChannel(4, 50)).toThrow(
    /invalid channel: 4. It must be a positive integer smaller than 3/,
  );
});

test('get channel', () => {
  const image = testUtils.createRgbaImage([
    [1, 2, 3, 1],
    [1, 2, 3, 1],
    [11, 2, 2, 2],
    [1, 3, 3, 3],
    [1, 6, 4, 3],
  ]);

  const expected = [
    [1, 1, 11, 1, 1],
    [2, 2, 2, 3, 6],
    [3, 3, 2, 3, 4],
    [1, 1, 2, 3, 3],
  ];
  for (let i = 0; i < image.channels; i++) {
    expect(image.getChannel(i)).toStrictEqual(expected[i]);
  }
});

test('fill alpha', () => {
  const img = new Image(1, 2, { colorModel: 'RGBA' });
  img.fillAlpha(0);
  expect(img).toMatchImageData([
    [0, 0, 0, 0],
    [0, 0, 0, 0],
  ]);
  img.fillAlpha(50);
  expect(img).toMatchImageData([
    [0, 0, 0, 50],
    [0, 0, 0, 50],
  ]);
});

test('fill alpha should throw if no alpha', () => {
  const img = new Image(1, 1);
  expect(() => img.fillAlpha(50)).toThrow(
    /fillAlpha can only be called if the image has an alpha channel/,
  );
});

test('check custom inspect, RGB image', () => {
  const image = new Image(1, 2);

  expect(inspect(image)).toMatchSnapshot();
});

test('check custom inspect, GREYA image', () => {
  const image = new Image(2, 2, { colorModel: 'GREYA' });

  expect(inspect(image)).toMatchSnapshot();
});

test('check custom inspect with image too large', () => {
  const image = new Image(25, 25, { colorModel: 'GREYA' });

  expect(inspect(image)).toMatchSnapshot();
});

test('check getColumn and getRow methods on an RGBA image', () => {
  const image = testUtils.createRgbaImage([
    [1, 2, 3, 1],
    [1, 2, 3, 1],
    [11, 2, 2, 2],
    [1, 3, 3, 3],
    [1, 6, 4, 3],
  ]);

  const resultRow = image.getRow(2);
  const resultColumn = image.getColumn(0);
  expect(resultRow).toStrictEqual([[11], [2], [2], [2]]);
  expect(resultColumn).toStrictEqual([
    [1, 1, 11, 1, 1],
    [2, 2, 2, 3, 6],
    [3, 3, 2, 3, 4],
    [1, 1, 2, 3, 3],
  ]);
});
test('check getColumn and getRow methods', () => {
  const image = testUtils.createGreyImage([
    [1, 5, 1],
    [1, 2, 1],
    [11, 2, 2],
  ]);

  const resultRow = image.getRow(2);
  const resultColumn = image.getColumn(0);
  expect(resultRow).toStrictEqual([[11, 2, 2]]);
  expect(resultColumn).toStrictEqual([[1, 1, 11]]);
});
