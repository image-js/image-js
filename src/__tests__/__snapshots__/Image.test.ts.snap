// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`check custom inspect with image too large 1`] = `
"Image {
  width: 25
  height: 25
  bitDepth: 8
  colorModel: GREYA
  channels: 2
  data: [...]
}"
`;

exports[`check custom inspect, GREYA image 1`] = `
"Image {
  width: 2
  height: 2
  bitDepth: 8
  colorModel: GREYA
  channels: 2
  data: {
    [
     [  0 255   0 255]
     [  0 255   0 255]
    ]
  }
}"
`;

exports[`check custom inspect, RGB image 1`] = `
"Image {
  width: 1
  height: 2
  bitDepth: 8
  colorModel: RGB
  channels: 3
  data: {
    [
     [  0   0   0]
     [  0   0   0]
    ]
  }
}"
`;
